### PR TITLE
Task1

### PR DESCRIPTION
**Pull request by Harshal662**
edited .ipynb

### CODE REVIEW

File: paperpi/library/Plugin.ipynb [hunk 1]
```diff
@@ -141,7 +144,131 @@
...
+    "        @property\n",
+    "        def update_function(self):\n",
+    "            return self._update_function\n",
+    "\n",
+    "        @update_function.setter\n",
+    "        def update_function(self, value):\n",
+    "            if value is not None and not callable(value):\n",
+    "                raise TypeError(\n",
+    "                    \"update_function should be a function or None.\")\n",
+    "            self._update_function = value\n",
```
Comment (txoof) [inline paperpi/library/Plugin.ipynb:187]: this duplicates the setter on line 249; I don't think this was intended.

File: paperpi/library/Plugin.ipynb [hunk 2]
```diff
@@ -141,7 +144,131 @@
     "        self.max_priority = max_priority\n",
     "        self.hash = self._generate_hash()\n",
     "        \n",
-    "    \n",
+    "        @property\n",
+    "        def resolution(self):\n",
+    "            return self._resolution\n",
+    "\n",
+    "        @resolution.setter\n",
```
Comment (txoof) [inline paperpi/library/Plugin.ipynb:151]: duplicates setter on line 230

File: paperpi/library/Plugin.ipynb [hunk 3]
```diff
@@ -77,6 +77,9 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "class PluginError(Exception):\n",
```
Comment (txoof) [inline paperpi/library/Plugin.ipynb:80]: I don't see where this is used. I like the idea of gathering all the Attribute/TypeErrors together under this exception class, but it's not used.

**Review by txoof — CHANGES_REQUESTED** [review]
several of the setters are duplicated this causes inconsistent output.

File: paperpi/library/Plugin.ipynb (hunk 1 above)
Comment (Harshal662) [reply paperpi/library/Plugin.ipynb:187]: Okay I will remove these lines

File: paperpi/library/Plugin.ipynb (hunk 2 above)
Comment (Harshal662) [reply paperpi/library/Plugin.ipynb:151]: will also remove these lines.

File: paperpi/library/Plugin.ipynb (hunk 3 above)
Comment (Harshal662) [reply paperpi/library/Plugin.ipynb:80]: Okay I will remove this class

**Review by txoof — CHANGES_REQUESTED** [review]
There are still some duplicated setter methods. The one that jumps out first is the layout setter; your change gets clobbered by the later.

It might be a good idea to do the following:

1. Group all the setter/getter methods below the functions for consistency
2. Pair all the getter/setters

**Review by txoof — CHANGES_REQUESTED** [review]
I'm sorry, I should have caught this sooner. There's a major problem with this PR: It appears that you've moved a bunch of the setters/getters into the `__init__` method. This means that they are **only** available when initing the object. 

I think this branch is too badly diverged from the main branch at this point to fix it up. It would be best to close this PR and try again. If you're up to it.

It would be good to test the plugin class against some "bad data" and double check that all the exceptions all work as expected before submitting the PR. 

here's an example:
```
p = Plugin(resolution=(640, 480), 
            name=None,
            layout=[],
            update_function=7,
            max_priority='a', # this will be accepted though I'm not entirely sure why.
#             refresh_rate=60,
#             min_display_time=30,
#             config={},
#             cache=None,
#             force_onebit=False,
#             screen_mode='1',
#             **kwargs
          )
```

**Pull request closed without merging**